### ISSUE TITLE
`glob` parameter description is misleading

### ISSUE THREAD

**Issue by Ps0ke** [smarie/python-pytest-cases#147]
The API Reference says about the [`glob` parameter][1]

 > - `glob`: an optional glob-like pattern for case ids, for example "*_success" or "*_failure". Note that this is applied on the case id, and therefore if it is customized through [`@case(id=...)`](#case) it should be taken into account.

Unfortunately it is not clear at all, what "glob-like" means.
The [man page for `bash`][2]

> Bash itself cannot recognize Regular Expressions. Inside scripts, it is commands and utilities -- such as sed and awk -- that interpret RE's.
> 
> Bash does carry out filename expansion [1] -- a process known as globbing -- but this does not use the standard RE set. Instead, globbing recognizes and expands wild cards. Globbing interprets the standard wild card characters [2] -- * and ?, character lists in square brackets, and certain other special characters (such as ^ for negating the sense of a match). There are important limitations on wild card characters in globbing, however. Strings containing * will not match filenames that start with a dot, as, for example, .bashrc. [3] Likewise, the ? has a different meaning in globbing than as part of an RE.

[Python's own `glob` implementation][3] has a slightly different flavor

> The glob module finds all the pathnames matching a specified pattern according to the rules used by the Unix shell, although results are returned in arbitrary order. No tilde expansion is done, but *, ?, and character ranges expressed with [] will be correctly matched. 

Note how there is no mention of negation by `^` or tilde expansion.

In order to figure out how to properly do negation in my `pytest` `glob` pattern, I had to go through the source code and found
https://github.com/smarie/python-pytest-cases/blob/d42b296c3ca0109dda9be71357213994ea911232/pytest_cases/case_parametrizer_new.py#L133

So actually the "glob-like" pattern is not a `glob` pattern, but a regular expression with `*` as a shorthand for `.*`. I personally find this really misleading. The `glob` parameter matches the case _id_. If I were to, for example, include a special character like `$` in my id, I would have to escape it in the regular expression, but I wouldn't know that I have to, because the fact that the `glob` pattern is a regular expression is hidden from me. Or, like in my case, I wanted to _exclude_ some pattern and tried matching for `^special_case*`. As the carret is listed in the bash docs as the pattern negation operator. But in regex it does the exact opposite, it matches the beginning of line, _including_ all the cases I wanted excluded and vice versa. What I needed to do is `(?!special_case)*`, which is not `glob`-like at all by any standard. One shouldn't need to read the source code to figure that out. This is clearly a documentation bug.

I would propose the following: Either

1. Trust developers that use `pytest_cases` to know regular expressions or be able to quickly read up for the specific needs. Rename the `glob` paramter to `match` (this might be problematic as it would break existing cases, maybe still accept `glob` as a fallback?) and remove the `*` → `.*` shorthand. In the documentation just say it excepts an regex that matches against the case id. It's not that hard to type the extra period and makes it easy to leverage the full power of regular expressions _that is available already anyway_.
2. Specify _exactly_ in the documentation what is meant by "glob-like" pattern. Choose a `glob` dialect and implement it properly, avoiding escaping of regex special characters that are not glob special characters. While this would keep the old API, it would still break existing code that relies on the regex behavior, all while being less powerfull and more work to implement correctly.
3. Just change the documentation to
> an optional regex-like pattern where `*` has to be used instead of `.*`. The use of a `*` repition quantifier in combination with an arbitrary character set `[…]` is impossible.

Clearly 2nd and 3rd are inferior solutions. 3rd does not fix the regex special character bug. Technically shell globbing wouldn't even match `special_case` on the pattern `special`, you'd have to use the pattern `*special*`. Yet the current implementation using regex will match that string. Clearly this is not consistent and it is so without a reason. Whats the rationale behind hiding regexes from developers?

_I hope this doesn't sound too condescending, I really enjoy using your library and appreciate the developers' work that went into it. I just want to help improving it :)_

[1]: https://github.com/smarie/python-pytest-cases/blob/04706868972a6b1a89b0de42d409c0bdbae75f05/docs/api_reference.md#L75
[2]: https://tldp.org/LDP/abs/html/globbingref.html
[3]: https://docs.python.org/3.8/library/glob.html

**Comment by smarie** [comment]
Thanks @Ps0ke for opening this issue, and for the detailed explanation and proposals ! 

Indeed the doc for API reference was not rendered properly: the stars were not escaped in the markdown.

I'll propose something resembling your proposal 3 for now, but simpler:

> `glob`: a matching pattern for case ids, for example `*_success` or `*_failure`. The only special character that can be used for now in this pattern is `*`, it can not be escaped, and it can be used several times in the same expression.

In parallel I'll modify the code so that all regex special characters are automatically escaped. That way, this parameter is definitely not a regex.

Now for your last question "what is the rationale hiding regexes from developers", I hesitated a long time before taking that decision, but it is motivated by simplicity: in most cases the case ids are python function names, which already do not have a wide character set, and in most use cases, people will wish to perform simple matches here. So a glob-like syntax for the joker char (* and not .*) is more appealing to me. I am a fervent user of regexp in many other contexts so I understand that this might become needed at some time. 

For this I think that the current workaround of passing a custom `filter` is already a valid path, did you try it ? Maybe it should be better documented ? Alternatively I am open to a third parameter named `regex` if `filter` is really too complex to use to work with regex.

**Comment by Ps0ke** [comment]
That sounds alright to me. Even though I'm not fully sold on the simplicity, I understand your rationale. I haven't looked using a `filter` as I just ended up expoliting the regex under the hood. But I kind of see now why you'd be inclined to offer a "basic filtering glob" and a "full powered callable" approach. I guess something like `filter=lambda f: re.search("(?!special_case)", f.__name__)` would be sufficient. Under that light, I'd support not including full regex support, provided it is implemented safely and the behaviour properly documented. :)

Concerning an additional parameter `regex`: I think when including a small example like the snippet I mentioned before in the documentation that is not needed :)

I actually might be interested in helping implementing and documenting the feature. Unfortunately, I'll be too busy until the holidays. Would you be interested in waiting until the beginning of next year for me to propose a pull request implementing and documenting proper regex escaping and glob syntax? I have never contributed to a Python project before and this seems like a manageable task to start with.

**Comment by smarie** [comment]
This is a great proposal @Ps0ke , a contribution would be very much appreciated ! Unfortunately I discovered your message a bit too late yesterday and had already made the mods in doc and tests. (I pushed it today in 2.5.0).

Maybe for another one, or for that "regex" additional parameter if you find it missing after all ?
Thanks a lot once again for opening this. Most people don't know it, but opening a relevant and well explained issue/feature request is half if not more of the job of contributing to open source code.
